_posts/2020-05-18-microservice-architectures.md

```diff
@@ -0,0 +1,80 @@
+
+Events force you to think about distributed computing - and what the **business process** should be in the event of failure: "what should happen if payment processing fails?".  This is inherent complexity in all systems, but provides a clear boundary around the domain when building [anti-fragile](https://refuses.github.io/preprints/antifragile.pdf) systems.
+
+The event schema, is the point of [coupling](https://www.slideshare.net/mtnygard/uncoupling-101026908):
+- The event schema is 'owned' by the producer.
+- Consumers and producers have weak operational coupling: both can be deployed independently, if the event schema is maintained.
+- Consumers have strong semantic coupling to the events schema they consume, rather than producers themselves.
+- Consumers/producers have weak development coupling: the implementation of either can change independently so long as the schema semantics is maintained.
+
+The [rivers-rapids pattern](https://www.youtube.com/watch?v=J3ihF11dpJY) enables service integration and avoids issues of [service discovery](https://microservices.io/patterns/server-side-discovery.html).  The hub/spoke model uses a single, highly available, event bus such as SNS.
+
+![rivers rapids pattern showing a single event bus: rapids; topic specific buses: rivers; and microservice consumers: ponds](/img/microservice-architectures/rivers-rapids.png)
+
+## Domain Oriented
+
+Event driven systems complement, and necessitate, [domain driven design](https://martinfowler.com/tags/domain%20driven%20design.html).  The focus on domain events identify natural service boundaries and decouple producers and consumers.  As a result, a microservice often comprises of a collection of highly cohesive, tightly coupled functions.  Using event boundaries, we are able to disregard the inherent complexity in the production of events the service consumes.  The generation of a [`CreateManagedAccount`](https://docs.aws.amazon.com/controltower/latest/userguide/lifecycle-events.html#create-managed-account) contains a large amount of complexity, orchestrating the creation of a new AWS account.  If you want to deploy a Config Rule, as part of your cloud governance strategy, to all new accounts you can ignore all that incidental complexity and simply respond to the event.  
```

Comment:
_Event driven systems complement and necessitate domain driven design._

Can lose the commas; the sentence is short enough.

Reply:
_The generation of a [`CreateManagedAccount`](https://docs.aws.amazon.com/controltower/latest/userguide/lifecycle-events.html#create-managed-account) *event abstracts the complexity of orchestrating the creation of a new AWS account*_

_If you want to deploy a Config Rule as part of your cloud governance strategy to all new accounts, you can ignore all that incidental complexity and simply respond to the event._